allennlp/modules/vision/region_detector.py

```diff
@@ -0,0 +1,137 @@
+            rpn_iou_labels=rpn_iou_labels,
+            rpn_batch_size_per_image=rpn_batch_size_per_image,
+            rpn_positive_fraction=rpn_positive_fraction,
+            rpn_bbox_reg_loss_type=rpn_bbox_reg_loss_type,
+            rpn_bbox_reg_loss_weight=rpn_bbox_reg_loss_weight,
+            rpn_bbox_reg_weights=rpn_bbox_reg_weights,
+            rpn_smooth_l1_beta=rpn_smooth_l1_beta,
+            rpn_loss_weight=rpn_loss_weight,
+            rpn_pre_nms_topk_train=rpn_pre_nms_topk_train,
+            rpn_pre_nms_topk_test=rpn_pre_nms_topk_test,
+            rpn_post_nms_topk_train=rpn_post_nms_topk_train,
+            rpn_post_nms_topk_test=rpn_post_nms_topk_test,
+            rpn_nms_thresh=rpn_nms_thresh,
+            rpn_bbox_loss_weight=rpn_bbox_loss_weight,
+            test_detections_per_image=detections_per_image,
+        )
+        pipeline = detectron.get_pipeline_from_flat_parameters(flat_parameters, make_copy=False)
+        self.model = pipeline.model
+        self.detections_per_image = detections_per_image
+
+    def forward(
+        self, raw_images: FloatTensor, image_sizes: IntTensor, featurized_images: FloatTensor
+    ) -> Dict[str, torch.Tensor]:
+        batch_size = len(image_sizes)
+        # RPN
+        from detectron2.structures import ImageList
+
+        image_list = ImageList(raw_images, [(image[0], image[1]) for image in image_sizes])
+        assert len(self.model.proposal_generator.in_features) == 1
+        featurized_images_in_dict = {
+            self.model.proposal_generator.in_features[0]: featurized_images
+        }
+
+        proposals, _ = self.model.proposal_generator(image_list, featurized_images_in_dict, None)
+
+        # this will concatenate the pooled_features from different images. 
+        _, pooled_features = self.model.roi_heads.get_roi_features(
+            featurized_images_in_dict, proposals
+        )
+        
+        predictions = self.model.roi_heads.box_predictor(pooled_features)
+        
+        # class probablity
+        cls_probs = [F.softmax(prediction, dim=-1) for prediction in predictions]
+        cls_probs = [cls_prob[:, :-1] for cls_prob in cls_probs] # background is last
+
+        predictions, r_indices = self.model.roi_heads.box_predictor.inference(
+            predictions, proposals
+        )
+        
+        features = []
+        proposal_start = 0
+        for image_num, image_proposal_indices in enumerate(r_indices):
+            feature_cell = {}
+            feature_cell['boxes'] = proposals[image_num].proposal_boxes.tensor[image_proposal_indices]
+            feature_cell['probs'] = cls_probs[image_num][image_proposal_indices]
+            feature_cell['features'] = torch.narrow(pooled_features, 0, proposal_start, len(proposals[image_num]))[image_proposal_indices]
+            proposal_start += len(proposals[image_num])
+            features.append(feature_cell)
+            
+        return features
```

Comment:
The output is supposed to be a `Dict[str, Tensor]`, but this output is a `List`.

Reply:
I guess that's more of a question for @jiasenlu. What happened here?